src/main/java/org/spongepowered/api/event/service/permission/ChangeSubjectEvent.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.service.permission;
+
+import org.spongepowered.api.service.context.Context;
+import org.spongepowered.api.service.permission.Subject;
+import org.spongepowered.api.service.permission.change.OptionChange;
+import org.spongepowered.api.service.permission.change.PermissionChange;
+
+import java.util.List;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * Events to report when changes occur to permissions subjects
+ */
+public interface ChangeSubjectEvent extends TargetSubjectEvent {
+
+    /**
+     * The set of {@link Context}s that this value was changed in
+     *
+     * @return The contexts changed
+     */
+    Set<Context> getContexts();
+
+    /**
+     * Event thrown when a subject has a permission change.
+     */
+    interface Permission extends ChangeSubjectEvent {
+
+        /**
+         * Get the list of changed permissions. This list is immutable.
+         *
+         * @return The changed permissions
+         */
+        List<PermissionChange> getChangedPermissions();
+
+        default void forEach(Consumer<PermissionChange> func) {
```

Comment:
If you name the method like this you cannot implement `ChangeSubjectEvent.Option` and `ChangeSubjectEvent.Permission` at the same time.
I'm not sure whether this is necessary, but it might be useful for bulk/transactional updates.